metricflow/time/time_granularity_solver.py

```diff
@@ -116,35 +128,75 @@ def resolve_granularity_for_partial_time_dimension_specs(
             )
+            for path_key in valid_group_by_elements.path_key_to_linkable_dimensions:
+                if (
+                    path_key.element_name == partial_time_dimension_spec.element_name
+                    and path_key.entity_links == partial_time_dimension_spec.entity_links
+                    and path_key.time_granularity is not None
+                ):
+                    minimum_time_granularity = (
+                        path_key.time_granularity
+                        if minimum_time_granularity is None
+                        else min(minimum_time_granularity, path_key.time_granularity)
+                    )
+            if not minimum_time_granularity:
+                raise RequestTimeGranularityException(
+                    f"Unable to resolve the time dimension spec for {partial_time_dimension_spec}. "
+                    f"Valid group by elements are:\n"
+                    f"{pformat_big_objects([spec.qualified_name for spec in valid_group_by_elements.as_spec_set.as_tuple])}"
+                )
+        else:
+            minimum_time_granularity = self.get_min_granularity_for_partial_time_dimension_without_metrics(
+                read_nodes=read_nodes,
+                node_output_resolver=node_output_resolver,
+                partial_time_dimension_spec=partial_time_dimension_spec,
+            )
+            if not minimum_time_granularity:
+                raise RequestTimeGranularityException(
+                    f"Unable to resolve the time dimension spec for {partial_time_dimension_spec}. "
+                )
 
         return minimum_time_granularity
 
+    def get_min_granularity_for_partial_time_dimension_without_metrics(
+        self,
+        read_nodes: Sequence[BaseOutput],
+        node_output_resolver: DataflowPlanNodeOutputDataSetResolver,
+        partial_time_dimension_spec: PartialTimeDimensionSpec,
+    ) -> Optional[TimeGranularity]:
+        """Find the minimum."""
+        granularity_free_qualified_name = StructuredLinkableSpecName(
+            entity_link_names=tuple(
+                [entity_link.element_name for entity_link in partial_time_dimension_spec.entity_links]
+            ),
+            element_name=partial_time_dimension_spec.element_name,
+        ).granularity_free_qualified_name
+        for read_node in read_nodes:
+            output_data_set = node_output_resolver.get_output_data_set(read_node)
+            for time_dimension_instance in output_data_set.instance_set.time_dimension_instances:
+                if time_dimension_instance.spec.date_part:
+                    continue
+                time_dim_name_without_granularity = StructuredLinkableSpecName.from_name(
+                    time_dimension_instance.spec.qualified_name
+                ).granularity_free_qualified_name
+                if time_dim_name_without_granularity == granularity_free_qualified_name:
+                    return time_dimension_instance.spec.time_granularity
```

Comment:
This is the minimum because it comes from the semantic manifest definition directly, is that right?

There's an assumption here that we'll never have duplicated read nodes or the ability to have a set of granularities rather than just one defined in a given semantic model, but I don't think that'll hold - at some point we will need derived semantic models or container time dimensions or something and this logic might not hold up.

For now - and this can be done in a follow-up - let's update this logic to collect all matches and assert that there's exactly one (or just return the minimum of that set).

Reply:
> This is the minimum because it comes from the semantic manifest definition directly, is that right?

Correct!

Sounds good I'll make that update.

Reply:
Actually, this behaves slightly different than I thought. The dataset has a time dimension for every supported granularity, so here I'll grab them all and take the minimum one.